api/genotypes_loader.api.inc

```diff
@@ -91,6 +109,110 @@ function genotypes_loader_helper_add_record_with_mode($record_type, $table, $mod
+ * @return
+ *   TRUE if success (meaning, we reached the end of the function, not neccesarily that everything went swimmingly), FALSE if failed.
+ */
+
+function genotypes_loader_helper_copy_records($record_count, $record_type, $table, $insert_values = array(), $file_name = NULL, $final_chunk = FALSE) {
+
+  // The number of records we want to reach before copying them in all at once
+  $copy_chunk_size = 10000;
+
+  // If we weren't given the name of the file to copy from, then try to find it using the record_type and table name
+  // that we were given
+  if ($file_name === NULL) {
+    $file_stub = '/tmp/genotype_loader.';
+    $record_type = str_replace(' ', '_', $record_type);
+    $record_type = strtolower($record_type);
+    $file_name = $file_stub . $table . '-' . $record_type . '.csv';
+  }
+
+  // If this is the last chunk to be entered into the database, then we don't have a record to add to the file.
+  // Otherwise, open the file to write to since this is needed by fputcsv.
+  // @TODO "Lacey: I'm concerned this might be a point of slowness. Perhaps there is a better
+  //       way to create CSV that doesn't require fputcsv and is more reliable
+  //       then simply imploding the values with comma's in-between."
+  if (!$final_chunk) {
+    // If we don't have a record to load in, something has defintiely gone wrong
+    if ($record_count == 0) {
+      return drush_set_error(dt('There is no genotype call to print to @filename', array('@filename' => $file_name)));
+    }
+    $file_handle = fopen($file_name, 'a') or die ("ERROR: Unable to create $file_name to bulk load records!\n");;
+    ksort($insert_values); // to ensure they're always in the same order.
+    fputcsv($file_handle, $insert_values);
+    fclose($file_handle);
+  }
+
+  if (($record_count%$copy_chunk_size == 0) || ($final_chunk)) {
+    // Write out the COPY command
+    // Columns are listed in alphabetical order because we used ksort() before printing the values to a file.
+    $copy_command = "COPY chado.genotype_call (
+      genotype_id,
+      marker_id,
+      meta_data,
+      project_id,
+      stock_id,
+      variant_id)
+      FROM STDIN WITH CSV";
+
+    genotypes_loader_remote_copy($copy_command, $file_name);
+
+    // Wipe the file clean by reopening it as write-only
+    $file_handle = fopen($file_name, 'w');
+    fclose($file_handle);
+
+    drush_log("Copied in " . $record_count . " records.");
+  }
+
+  return TRUE;
+
+}
+
+/**
+ * Uses drush to safely copy data quickly. ONLY USE IN TRIPAL JOBS!!
+ *
+ * PostgreSQL COPY is extremely effective at copying/inserting large amount of
+ * data. This modules uses it to create materialized views to cache queries
+ * for faster data access. However, using chado_query('COPY...') poses many
+ * security risks in a web setting. We use a combination of drush sql-cli,
+ * which opens a psql session, and psql's \copy command to take advantage of
+ * the fact that Tripal Jobs are run on the command-line.
+ *
```

Comment:
This needs to be re-worded as it's currently specific to nd_genotypes. Perhaps

PostgreSQL COPY is extremely effective at copying/inserting large amount of data. This modules uses it to **make inserting massive genotype files more efficient**. However, using chado_query('COPY...') poses many security risks in a web setting. **Instead we use Drupal to determine the psql connection string,** and psql's \copy command to take advantage of the fact that Tripal Jobs are run on the command-line.

Note: Bolded sections indicate my changes.

Reply:
Also addressed in commit https://github.com/UofS-Pulse-Binfo/genotypes_loader/commit/1277bb2ef06c2db55a61777187950a4e50ff4e10